New resource `azurerm_policy_remediation`

ArcturusZhang opened the pull request:
This is a revised version of PR [5277](https://github.com/terraform-providers/terraform-provider-azurerm/pull/5277).

A `azurerm_policy_remediation` is used along with a `azurerm_policy_assignment` to remediate a policy.
A policy can be assigned in 4 different scopes: Subscription, Resource Group, Resource and Management Group. This remediation resource supports all of them using a `scope` field similar to the `azurerm_policy_assignment`.

Since the id of a management group (like `/providers/Microsoft.Management/managementGroups/00000000-0000-0000-0000-000000000000`) is not a valid azure resource id, passing it to `azure.ParseResourceId` will throw errors, I wrote some additional ID parsing functions to deal with this kind of situations.

And additionally, since all of the ids in a remediation will be converted to lower cases, I am using case diff suppression everywhere.

katbyte commented on azurerm/internal/services/policyinsights/tests/resource_arm_policy_remediation_test.go:
```diff
@@ -0,0 +1,591 @@
+
+func testAccAzureRMPolicyRemediation_atSubscription(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+data "azurerm_subscription" "current" {}
```
We'll need to add a provider block to all these tests

ArcturusZhang replied:
Done

katbyte commented on azurerm/internal/services/policyinsights/resource_arm_policy_remediation.go:
```diff
@@ -0,0 +1,207 @@
+				},
+			},
+
+			"policy_definition_reference_id": {
```
I don't see any test that make use of this property?

ArcturusZhang replied:
I will add some tests for this property

ArcturusZhang replied:
Test added

katbyte commented on azurerm/internal/services/policyinsights/parse/remediation.go:
```diff
@@ -0,0 +1,130 @@
+	segments := regex.Split(input, -1)
+
+	if len(segments) != 2 {
+		return nil, fmt.Errorf("Unable to parse Policy Remediation ID %q: Expected 2 segments after splition", input)
```
```suggestion
		return nil, fmt.Errorf("Unable to parse Policy Remediation ID %q: Expected 2 segments after split", input)
```

katbyte commented on azurerm/internal/services/policyinsights/parse/remediation.go:
```diff
@@ -0,0 +1,130 @@
+	return &id, nil
+}
+
+type RemediationScopeId struct {
```
I'm not sure we should share the properties struct like this. not all properties will be populated for each of the remediation types. While they share a similar base we should have 4 derived types for each type of ID.eg PolicyRemediationAtSubscriptionID` and such

ArcturusZhang replied:
I have done some tweaks on the structure of this part. Introduced an interface for `RemediationScopeId` and 4 structs implemented this interface, please have a look again

katbyte commented on azurerm/internal/services/policyinsights/parse/remediation.go:
```diff
@@ -0,0 +1,130 @@
+	return err == nil
+}
+
+// TODO -- move this to management group RP directory
```
Could we do this as part of this PR & update the resources to use it?

ArcturusZhang replied:
Well, I have already opened a PR for this (and it has been reviewed). Could I just close that PR, and merge those changes into this PR?

ArcturusZhang replied:
I have merged [PR 5878](https://github.com/terraform-providers/terraform-provider-azurerm/pull/5878) into this PR.

katbyte replied:
awesome, thanks 👍

katbyte commented on azurerm/internal/services/policyinsights/remediations.go:
```diff
@@ -0,0 +1,57 @@
+	default:
+		return policyinsights.Remediation{}, fmt.Errorf("Error creating Policy Remediation: Invalid scope type %q", scope.Type)
+	}
+}
```
Given this will only be called in one resource do we need to have a function for it?

ArcturusZhang replied:
Sure, i will move these functions to the corresponding CRUD functions

ArcturusZhang replied:
Since the GetAtScope function and DeleteAtScope function are called multiple times, I will keep them, but move them to the resource file.

ArcturusZhang commented on azurerm/internal/clients/client.go:
```diff
@@ -46,6 +46,7 @@ import (
 	network "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/network/client"
 	notificationhub "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/notificationhub/client"
 	policy "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/policy/client"
+	policyinsights "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/policyinsights/client"
```
Sure, I will move them to the policy directory

ArcturusZhang commented on azurerm/internal/services/policyinsights/client/client.go:
```diff
@@ -0,0 +1,19 @@
+package client
```
Done

ArcturusZhang commented on azurerm/internal/services/policyinsights/data_source_policy_remediation.go:
```diff
@@ -0,0 +1,68 @@
+package policyinsights
```
You are right, this resource is like policy assignment. The data source does not make much sense. I have removed it.

ArcturusZhang commented on azurerm/internal/services/policyinsights/data_source_policy_remediation.go:
```diff
@@ -0,0 +1,68 @@
+		}
+		return fmt.Errorf("Error reading Policy Remediation %q (Scope %q): %+v", name, scope, err)
+	}
+	if resp.ID == nil || *resp.ID == "" {
```
The data source has been removed

ArcturusZhang commented on azurerm/internal/services/policyinsights/data_source_policy_remediation.go:
```diff
@@ -0,0 +1,68 @@
+	if resp.ID == nil || *resp.ID == "" {
+		return fmt.Errorf("Cannot read Policy Remediation %q (Scope %q) ID", name, scope)
+	}
+
```
The data source has been removed

ArcturusZhang commented on azurerm/internal/services/policyinsights/parse/remediation.go:
```diff
@@ -0,0 +1,130 @@
+	RemediationScopeId
+}
+
+// TODO: This paring function is currently suppressing every case difference due to github issue: https://github.com/Azure/azure-rest-api-specs/issues/8353.
```
Done

ArcturusZhang commented:
Hi @katbyte I have resolved your comment. These are what I have done:
1. Removed directory and package `policyinsights` and move all the files to `policy` package.
1. Removed the data source of `azurerm_policy_remediation`
1. Define an interface `PolicyScopeId` to represent different scope IDs, and declare 4 different structs for 4 different scopes implementing the interface `PolicyScopeId`. These types could be used for other resources in the `azurerm_policy_assignment`, `azurerm_policy_definition` and `azurerm_policy_set_definition`.
1. Refactor the file structure in `parse` and `validate`. Since the scope can be shared within all the resources in `policy` service, I moved the interface`PolicyScopeId` and 4 structs for different scope IDs to the `parse/scope.go` file.

Despite I introduce the parsing functions for `azurerm_policy_definition` and `azurerm_policy_assignment`, I did not change the resource and data source of policy definition and policy assignment for simplicity of this PR. I could create another PR to refine the existing resource in policy package.

ArcturusZhang commented:
HI @katbyte some information on the acceptance tests of this resource:
`TestAccAzureRMPolicyRemediation_atSubscription`: PASS
`TestAccAzureRMPolicyRemediation_atResourceGroup`: PASS
`TestAccAzureRMPolicyRemediation_atManagementGroup`: FAIL, the service principal i am using does not have the permission to create management group
`TestAccAzureRMPolicyRemediation_atResource`: PASS
`TestAccAzureRMPolicyRemediation_atSubscriptionWithDefinitionSet`: FAIL, something wrong with the `azurerm_policy_set_definition` resource, it will always return diff in the `metadata` field with something like `createdOn`.

katbyte commented:
I think this will conflict with the prs: #5981 #6107 ? 🤔

ArcturusZhang commented:
> I think this will conflict with the prs: #5981 #6107 ? 🤔

Well, in general this PR should not have conflicts with them, because I did not change anything else in the policy service (including the policy_definition resource) other than the addition of policy remediation. Therefore I suppose my PR should not have conflicts with them.

But I think the changes for management group IDs in this PR would benefit those two PR. They are trying to handle the management group in their own resources, which I think is not a very good idea to introduce too much copy and paste work.

ArcturusZhang commented:
I just noticed that one of them has been approved, then maybe a refinement of how to deal the management group in the policy service should be introduced after that and this PR get merged

katbyte commented on azurerm/internal/services/policy/resource_arm_policy_remediation.go:
```diff
@@ -0,0 +1,248 @@
+}
+
+// RemediationGetAtScope is a wrapper of the 4 Get functions on RemediationsClient, combining them into one to simplify code.
+func RemediationGetAtScope(ctx context.Context, client *policyinsights.RemediationsClient, name string, scopeId parse.PolicyScopeId) (policyinsights.Remediation, error) {
```
Should this be private to this package?
```suggestion
func remediationGetAtScope(ctx context.Context, client *policyinsights.RemediationsClient, name string, scopeId parse.PolicyScopeId) (policyinsights.Remediation, error) {
```

ArcturusZhang replied:
Emmm I am afraid that we cannot do this... because this is also used in the acc tests, the checkDestroy or checkExist functions are using this to query whether the remediation is successfully provisioned.

katbyte replied:
👍

katbyte commented on azurerm/internal/services/policy/tests/resource_arm_policy_remediation_test.go:
```diff
@@ -0,0 +1,733 @@
+		}
+
+		client := acceptance.AzureProvider.Meta().(*clients.Client).Policy.RemediationsClient
+		ctx := acceptance.AzureProvider.Meta().(*clients.Client).StopContext
```
could we move these to the top of the anon function?

ArcturusZhang replied:
Sure!

katbyte commented on website/docs/r/policy_remediation.html.markdown:
```diff
@@ -0,0 +1,109 @@
+
+* `name` - (Required) The name of the Policy Remediation. Changing this forces a new resource to be created.
+
+* `scope` - (Required) The Scope at which the Policy Remediation should be applied, which must be a Resource ID (such as Subscription e.g. `/subscriptions/00000000-0000-0000-0000-000000000000` or a Resource Group e.g.`/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myResourceGroup` or a specified Resource e.g. `/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/ ...
```
this might read better as a list?

ArcturusZhang replied:
Sure

katbyte commented on website/docs/r/policy_remediation.html.markdown:
```diff
@@ -0,0 +1,109 @@
+The following attributes are exported:
+
+* `id` - The ID of the Policy Remediation.
+
```
could we add the timeout docs here?

ArcturusZhang replied:
Oops, I forgot this. I will add them

robinsk commented:
Nice to see getting closer to merge. Really looking forward to clean up our subscriptions programmatically.

@ArcturusZhang Thank you for working on this, much appreciated!

ArcturusZhang commented on azurerm/internal/services/managementgroup/data_source_management_group.go:
```diff
@@ -21,8 +22,9 @@ func dataSourceArmManagementGroup() *schema.Resource {
-				Required: true,
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validate.ManagementGroupName,
```
Here group_id means this one in its ID:
```
/providers/Microsoft.ManagementGroup/managementGroups/{group_id}
```
the thing that in other resources' IDs at the same position would be called as a `name`, for instance
```
/subscriptions/{subsID}/resourceGroups/group1/providers/Microsoft.Compute/virtualMachines/{name}
```
And another reason is that if we call this `group_id` in the validation function, the validation function would looks like `ManagementGroupGroupID` or `ManagementGroupID`. This would be confusing with the validation function of the real management group ID. Therefore I took this name.

ArcturusZhang commented on azurerm/internal/services/managementgroup/resource_arm_management_group.go:
```diff
@@ -38,10 +40,11 @@ func resourceArmManagementGroup() *schema.Resource {
+				Optional:     true,
+				Computed:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.ManagementGroupName,
```
Same reason as the above comment

ArcturusZhang commented on azurerm/internal/services/policy/parse/definition.go:
```diff
@@ -0,0 +1,43 @@
+// TODO: This paring function is currently suppressing every case difference due to github issue: https://github.com/Azure/azure-rest-api-specs/issues/8353
+func PolicyDefinitionID(input string) (*PolicyDefinitionId, error) {
+	// in general, the id of a definition should be:
+	// {scope}/providers/Microsoft.PolicyInsights/remediations/{name}
```
Yeah... This regex only matches the middle part, and then I did a split on the ID to separate the `{scope}` part and the `name`. Because the `{scope}` can either be a subscription ID, resource group ID, resource ID, or management group ID, those 4 scopes do not share the same pattern.

katbyte replied:
kk 👍

ArcturusZhang commented:
Hi @katbyte I have resolved the comments.

And I just spotted a typo in the doc of `azurerm_policy_set_definition` when I am copying the `timeout` section, and I changed those in this PR, hope you would not mind.

katbyte commented on azurerm/internal/services/policy/resource_arm_policy_remediation.go:
```diff
@@ -0,0 +1,248 @@
+	}
+
+	return utils.FlattenStringSlice(input.Locations)
+}
```
i think for both these functions we could just inline the expand/flatten slice functions?

ArcturusZhang replied:
Emm, in my memory, the inline call was the first version of this, and then I changed them into this... Let me check if there is problems when changing these back

ArcturusZhang replied:
Done

katbyte commented on azurerm/internal/services/policy/validate/definition.go:
```diff
@@ -0,0 +1,41 @@
+	scope := segments[0]
+	// scope should be a resource group ID, or Management Group ID
+	_, err := parse.PolicyScopeID(scope)
+	if err != nil {
```
these two lines could be merged

ArcturusZhang replied:
Done

ArcturusZhang commented:
Hi @katbyte I resolved the final two minor comments

ArcturusZhang commented:
I also included some improvement of policy_definition in this PR. Acc test result for the policy definition:
![image](https://user-images.githubusercontent.com/10554446/77717736-2a579700-701c-11ea-849b-1cd85cb6caf5.png)
But still the test case related with mgmt group fails with authorization error:
```
Status=403 Code="AuthorizationFailed" Message="The client '***' with object id '***' does not have authorization to perform action 'Microsoft.Management/managementGroups/read' over scope '/providers/Microsoft.Management/managementGroups/***' or the scope is invalid. If access was recently granted, please refresh your credentials."
```

kailunshi commented:
Hi @ArcturusZhang 

First at all, thank you for the great work. I know this is merged and you might not be following it but I discovered that ResourceDiscoveryMode option is not included. https://github.com/Azure/azure-sdk-for-go/blob/master/services/policyinsights/mgmt/2019-10-01/policyinsights/models.go#L2056

Could you please shed some light? let me know if you want me to create a new ticket.

ArcturusZhang commented:
> Hi @ArcturusZhang
> 
> First at all, thank you for the great work. I know this is merged and you might not be following it but I discovered that ResourceDiscoveryMode option is not included. https://github.com/Azure/azure-sdk-for-go/blob/master/services/policyinsights/mgmt/2019-10-01/policyinsights/models.go#L2056
> 
> Could you please shed some light? let me know if you want me to create a new ticket.

Well, this field initially does not exist in the version of go SDK terraform is using when I first work on this resource (Dec. 2019). I will take some time to add this into this resource, thanks for remind me.

kailunshi commented:
> > Hi @ArcturusZhang
> > First at all, thank you for the great work. I know this is merged and you might not be following it but I discovered that ResourceDiscoveryMode option is not included. https://github.com/Azure/azure-sdk-for-go/blob/master/services/policyinsights/mgmt/2019-10-01/policyinsights/models.go#L2056
> > Could you please shed some light? let me know if you want me to create a new ticket.
> 
> Well, this field initially does not exist in the version of go SDK terraform is using when I first work on this resource (Dec. 2019). I will take some time to add this into this resource, thanks for remind me.

Makes total sense. azure is fast evolving. :)

Thank you!
